File: algorithmic_toolbox/week4_divide_and_conquer/1_binary_search/cpp/binary_search.cpp
#include <iostream>
#include <vector>

using std::vector;

int binary_search(const vector<int> &a, int x) {
  int left = 0, right = (int)a.size() - 1;

  while (left <= right) {
    int mid = left + (right - left) / 2;

    if (a[mid] == x)
      return mid;

    if (a[mid] < x)
      left = mid + 1;
    else
      right = mid - 1;
  }

  return -1;
}

int linear_search(const vector<int> &a, int x) {
  for (size_t i = 0; i < a.size(); ++i)
    if (a[i] == x) return i;

  return -1;
}

File: algorithmic_toolbox/week4_divide_and_conquer/1_binary_search/python/binary_search.py
import sys


def binary_search(a: list[int], x: int) -> int:  # noqa: C901
    left, right = 0, len(a) - 1

    while left <= right:
        mid = (left + right) // 2

        if a[mid] == x:
            return mid

        if a[mid] < x:
            left = mid + 1
        else:
            right = mid - 1

    return -1


def linear_search(a: list[int], x: int) -> int:
    for i in range(len(a)):
        if a[i] == x:
            return i
    return -1


if __name__ == "__main__":
    data = list(map(int, sys.stdin.read().split()))
    n = data[0]
    m = data[n + 1]
    a = data[1 : n + 1]
    for x in data[n + 2 :]:
        print(binary_search(a, x), end=" ")
